### DISCUSSION TITLE
How to do cleanup in a workload React component

### DISCUSSION THREAD

**Discussion by box-turtle** [microsoft/Microsoft-Fabric-workload-development-sample#23] [Help] [answered]
Our workload is a notebook experience so when it loads we have to allocate some resources that need to be cleaned up if the user exits the page.

When the user leaves the page, we would expect the react component to unload and any cleanup methods from `useEffect` hooks or `windowbeforeunload` even handlers would fire.

However it seems that if the user navigates using buttons/links OUTSIDE the iframe, all this does is set `display:none` on the iframe container, so I can find no event or callback that gets fired in this case.

<img width="863" alt="image" src="https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/3999743/8014d260-50d2-4056-b4d5-413c53cc0c57">

Does the workloadClient expose any events when this kind of external navigation occurs? 

If not can you suggest any other way to handle cleanup tasks when the user navigates away?

**Comment by anitamayorov** [comment, marked as answer]
Hi @box-turtle,

You need to register to OnAfterNavigateAway event - you can take a look at the following sample code: https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/b64b0bac0b0f0f1433df9e26bde0839174cc796b/Frontend/src/controller/SampleWorkloadController.ts#L195

When catching that event, you can do all the desired cleanup.

**Comment by box-turtle** [comment]
Thanks Anita - that works perfectly.
